lisa/analysis/frequency.py

```diff
@@ -41,6 +42,47 @@ class FrequencyAnalysis(TraceAnalysisBase):
     def __init__(self, trace):
         super(FrequencyAnalysis, self).__init__(trace)
 
+    @requires_events('cpu_frequency')
+    def _check_freq_domain_coherency(self, cpus=None):
+        """
+        Check that all CPUs of a given frequency domain have the same frequency
+        transitions.
+
+        :param cpus: CPUs to take into account. All other CPUs are ignored.
+            If `None`, all CPUs will be checked.
+        :type cpus: list(int) or None
+        """
+        domains = self.trace.plat_info['freq-domains']
+        if cpus is None:
+            cpus = list(itertools.chain.from_iterable(domains))
+
+        if len(cpus) < 2:
+            return
+
+        df = self.trace.df_events('cpu_frequency')
+
+        for domain in domains:
+            # restrict the domain to what we care. Other CPUs may have garbage
+            # data, but the caller is not going to look at it anyway.
+            domain = set(domain) - set(cpus)
+            if len(domain) < 2:
+                continue
+
+            # Get the frequency column for each CPU in the domain
+            freq_columns = [
+                # drop the index since we only care about the transitions, and
+                # not when they happened
+                df[df['cpu'] == cpu]['frequency'].reset_index(drop=True)
+                for cpu in domain
+            ]
+
+            # Check that all columns are equal. If they are not, that means that
+            # at least one CPU has a frequency transition that is different
+            # from another one in the same domain, which is highly suspicious
```

Comment:
Good point, we never hit that since the current check is not more robust than this one against that. Shifting by one would work in case of comparison failure.